### ISSUE TITLE
API Proposal: TaskScheduler.Queue(Action, object)

### ISSUE THREAD

**Issue by benaadams** [dotnet/corefx#39994]
To support not creating an additional Task when the scheduler implements the method
```csharp
protected internal virtual void QueueTask (Action<object?> continuation, object? state)
    => Task.Factory.StartNew(
        continuation, 
        state, 
        CancellationToken.None, 
        TaskCreationOptions.DenyChildAttach, 
        this);
```
A place this that could be used is in [ManualResetValueTaskSourceCore](https://github.com/dotnet/corefx/blob/8a958aa3d1cf2e91f6e9cf839e023ad4f8d489f2/src/Common/src/CoreLib/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs#L185)
```csharp
case TaskScheduler ts:
    Task.Factory.StartNew(continuation, state, CancellationToken.None, TaskCreationOptions.DenyChildAttach, ts);
    break;
```
Where it could instead call
```csharp
case TaskScheduler ts:
    ts.Queue(continuation, state);
    break;
```

/cc @ReubenBond @stephentoub

**Comment by ReubenBond** [comment]
Other perf-sensitive areas where this API may be beneficial:

* `System.Threading.Channels`, eg: [`AsyncOperation.OnCompleted(Action<object>, object, short, ValueTaskSourceOnCompletedFlags)`](https://github.com/dotnet/corefx/blob/4e7d0c32d9cedc9d1436c3d71eec8ac96ed0064d/src/System.Threading.Channels/src/System/Threading/Channels/AsyncOperation.cs#L297) & [`AsyncOperation.SignalCompletion()`](https://github.com/dotnet/corefx/blob/4e7d0c32d9cedc9d1436c3d71eec8ac96ed0064d/src/System.Threading.Channels/src/System/Threading/Channels/AsyncOperation.cs#L409)

* [`System.Threading.Tasks.TaskSchedulerAwaitTaskContinuation`](https://github.com/dotnet/corefx/blob/4e7d0c32d9cedc9d1436c3d71eec8ac96ed0064d/src/Common/src/CoreLib/System/Threading/Tasks/TaskContinuation.cs#L492): this is less straightforward, but I believe we could avoid the additional `Task` allocation in most/all cases, though perhaps the proposed `QueueTask` method would need some 'can inline' flag or a `TryExecuteTaskInline` overload)

**Comment by stephentoub** [comment]
All of this assumes the TaskScheduler isn't going to allocate in order to store the arbitrary action/state pair.  I expect there are relatively few TaskScheduler implementations where that's the case.  Even the default TaskScheduler needs to allocate for that case (we explored using a queue of struct-based tuples instead in order to handle this possibility, but it made most operations slower).

**Comment by stephentoub** [comment]
Exception-related behavior is also an issue here.  The semantics of exceptions with tasks is clear.  It's much less clear with QueueTask as highlighted.   Presumably the implementation can't just be what's written above, as an arbitrary Action that throws is going to have its exception eaten.  Which then suggests that the base implementation as shown above is actually going to need to be more complicated to deal with that, to the point where I think the base implementation above will actually incur more allocation.  Can you elaborate?

**Comment by ReubenBond** [comment]
Regarding the exception case, I imagine we could follow similar semantics to ThreadPool.QueueUserWorkItem (i.e, bubble up to a global handler)

> we explored using a queue of struct-based tuples instead in order to handle this possibility, but it made most operations slower

I see your point, and any internal representation would have to go through much of the work which Task already implements (though presumably without the bits required to store a result or exception or signal completion).

Maybe https://github.com/dotnet/corefx/issues/37755 is more realistic. The crux of these asks is finding ways to reducing the cost of using custom TaskScheduler implementations in .NET.

**Comment by stephentoub** [comment]
> I imagine we could follow similar semantics to ThreadPool.QueueUserWorkItem (i.e, bubble up to a global handler)

Right.  Can you write out how the base TaskScheduler.QueueAction does that?  How does it do so without allocating?

**Comment by ReubenBond** [comment]
> Can you write out how the base TaskScheduler.QueueAction does that?

The base `TaskScheduler` wouldn't avoid allocations, it would behave similarly to `Task.Factory.StartNew`. The purpose of the proposal is to allow custom implementations to avoid work. For example, in our case, if we could set the current TaskScheduler (so that it's not `TaskScheduler.Default` during the invocation) then we can likely inline the call.

In the common case (where the ActivationTaskScheduler isn't currently executing on some thread) we could inline, following this flow:

```
set TaskScheduler.Current = this
invoke action(state)
reset TaskScheduler.Current
```

In the case where the ActivationTaskScheduler is currently executing already, we would have to enqueue the work item.

As for how exceptions in the callback are bubbled up, I think that would require an additional API. Each custom TaskScheduler could expose their own, but that would not be ideal.

**Comment by stephentoub** [comment]
> The base TaskScheduler wouldn't avoid allocations, it would behave similarly to Task.Factory.StartNew. The purpose of the proposal is to allow custom implementations to avoid work. For example, in our case, if we could set the current TaskScheduler (so that tasks started during the invocation are routed to the right scheduler) then we can likely inline the call.

My point is, to take advantage of this API, call sites like those you mention would need to switch to use them.  And unless the TaskScheduler overrode the method to be more efficient, the base implementaiton would actually end up being *worse* than the costs incurred today.

For example, it couldn't just be implemented as:
```C#
Task.Factory.StartNew(action, state, ...);
```
To deal with exceptions, it would need to be something more like:
```C#
var stateWrapper = new StateWrapper(action, state);
Task.Factory.StartNew(s =>
{
    var sw = (StateWrapper)s;
    try { sw.Action(sw.State); }
    catch (Exception e) { ... }
}, stateWrapper, ...);
```
so now we're not only incurring the Task allocation, we're also incurring a StateWrapper allocation.

**Comment by ReubenBond** [comment]
The current call sites (eg, ManualResetValueTaskSource) schedule using `Task.Factory.StartNew(action, state, ...)` without observing the resulting Task, presumably relying on `TaskScheduler.UnobservedTaskException` for the cases where `action(state)` throws. Implementations which inlined would need to come up with another way to bubble up the unhandled exception. Having a mechanism on `TaskScheduler` allowing implementations to fire that event seems feasible. My understanding of the problem is likely still lacking.

**Comment by stephentoub** [comment]
> presumably relying on TaskScheduler.UnobservedTaskException for the cases where action(state) throws

No, they rely on the fact that they're all used in contexts of async/await Action callbacks that should never throw (if they do, there's a much more fundamental bug).

**Comment by ReubenBond** [comment]
I see - thanks for clearing up my misunderstanding. Perhaps an API explicitly intended for the callbacks case would be viable? Or possibly a `TryExecuteInline(Action<object>, object)` API?

I hope there's room to improve the custom TaskScheduler case and enable some of the kinds of optimizations which the TaskScheduler.Default case has.
